src/encoder.js

```diff
@@ -270,41 +271,98 @@ export default class Encoder extends Extendable {
             }
-            let furtherEnabledExtensions = applyRecursively
-              ? [...enabledExtensions, extensionPoint]
-              : enabledExtensions;
-            this.encodeValue(extension.serialise(placeholder.value), extTarget, furtherEnabledExtensions);
-          } else {
-            this.encodeValue(placeholder.value, extTarget, enabledExtensions);
+            replacements.push({ index: placeholder.location, data: extTarget, placeholders: this.placeholderMap });
           }
-          target.splice(target.indexOf(placeholder), 1, ...extTarget);
         }
       }
 
-      if (typeof extension.memo === 'function') {
-        target.unshift(...this.encodeValue(extension.memo(), [], enabledExtensions));
+      this.placeholderMap = {};
+
+      let returnData = [];
+      let lastIndex = 0;
+
+      // restructure data into new array
+      for (let replacement of replacements) {
+        if (replacement.index < lastIndex) {
+          throw new Error('Corrupt non-linear state');
+        } else if (replacement.index > lastIndex) {
+          // skips placeholder byte
+          returnData.push(...target.slice(lastIndex, replacement.index));
+        }
+        for (let extensionPoint in replacement.placeholders) {
+          if (!(extensionPoint in this.placeholderMap)) {
+            this.placeholderMap[extensionPoint] = [];
+          }
+          let placeholders = replacement.placeholders[extensionPoint];
+          for (let i = 0; i < placeholders.length; ++i) {
+            let placeholder = placeholders[i];
+            this.placeholderMap[extensionPoint].push({ value: placeholder.value, extensionPoint: extensionPoint, location: placeholder.location + returnData.length });
+          }
+        }
+
+        // fill in replacement data
+        returnData.push(...replacement.data);
+        lastIndex = replacement.index + 1;
       }
+      
+      returnData.push(...target.slice(lastIndex));
+      target = returnData;
     }
 
+    if (includeMemos) { // build memos in series
+      let enabledExtensions = extensions.slice();
+      let prepend = [];
+      
+      while (enabledExtensions.length > 0) {
+        let extensionPoint = enabledExtensions.pop();
+        let extension = this.extensions[extensionPoint];
+        if (typeof extension.memo === 'function') {
+          this.placeholderMap = {};
+          let extTarget = [];
+          let predata = this.encodeValue(extension.memo(), extTarget, enabledExtensions);
+          let memoData = this.replaceExtensionPlaceholders(predata, enabledExtensions, false);
+          prepend.push(memoData);
+        }
+      }
+      // combine them all together
+      let finalData = [];
+      for (let i = prepend.length - 1; i >= 0; --i) {
+        finalData.push(...prepend[i]);
+      }
+      finalData.push(...((target : any) : SuperPackedValue));
+      return finalData;
+    }
     return ((target : any) : SuperPackedValue);
```

Comment:
Instead of casting `target`, we should be able to keep a separate array that only ever has bytes in it, and use that as the real output array, right?